File: java/src/main/java/fr/automated/trading/systems/portfoliosmanager/PricesRecord.java
package fr.automated.trading.systems.portfoliosmanager;

import fr.automated.trading.systems.utils.utils.AtsLogger;

import java.util.ArrayList;
import java.util.List;

@Deprecated
public class PricesRecord {

    private final List<Double> prices = new ArrayList<>();
	private final List<TradeRecords> observers = new ArrayList<>();
	
	public void addPrice(double price) {
		AtsLogger.log("Prices records updated");
		prices.add(price);
		notifyObservers();
	}
	
	public List<Double> getPrices() {
		return prices;
	}
	
	public double getPrice(int i) {
		return prices.get(i);
	}

	public double getLastPrice() {
		return prices.get(prices.size()-1);
	}
	
	public double getPreviousPrice() {
		return prices.get(prices.size()-2);
	}
	
	public void registerObserver(TradeRecords observer) {
		AtsLogger.log("Observer added");
		observers.add(observer);
	}
	
	public void notifyObservers() {
		AtsLogger.log("Notify Observers");
        for (TradeRecords observer : observers) {
            observer.update(getLastPrice());
        }
	}
	
	public void debug() {
	    for(double price : prices) {
            AtsLogger.log(price);
        }

    }
}

File: cpp/src/PricesRecord.cpp
#include "../include/PricesRecord.h"

PricesRecord::PricesRecord()
{
    _prices = new vector<double>;
    _observers = new vector<Record>;
    _count = 0;
}

PricesRecord::~PricesRecord()
{
    delete(_prices);
    delete(_observers);
}

void PricesRecord::addPrice(double price)
{
    _prices->push_back(price);
    _count++;
    notifyObservers();
}

vector<double> * PricesRecord::getPrices()
{
    return _prices;
}

double PricesRecord::getPrice(int id)
{
    return _prices->at(id);
}

int PricesRecord::getCount()
{
    return _count;
}

double PricesRecord::getLastPrice()
{
    return _prices->at(_count-1);
}

double PricesRecord::getPreviousPrice()
{
    return _prices->at(_count-2);
}

void PricesRecord::registerObserver(Record* observer)
{
    _observers->push_back(*observer);
}

void PricesRecord::notifyObservers()
{
    for(int i=0; i<_observers->size(); i++)
        _observers->at(i).update(getLastPrice());
}

void PricesRecord::debug()
{
    for(size_t i=0; i<_prices->size(); i++)
    {
        cout << "price[" << i << "] = " << _prices->at(i) << endl;
    }
}
